File: LabPA/Lab5/java/task-1/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Scanner;

public class Main {
    static class Task {
        public final static String INPUT_FILE = "in";
        public final static String OUTPUT_FILE = "out";

        static int n, k;

        private void readInput() {
            try {
                Scanner sc = new Scanner(new File(INPUT_FILE));
                n = sc.nextInt();
                k = sc.nextInt();
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(ArrayList<ArrayList<Integer>> result) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%d\n", result.size());
                for (ArrayList<Integer> arr : result) {
                    for (int i = 0; i < arr.size(); i++) {
                        pw.printf("%d%c", arr.get(i), i + 1 == arr.size() ?
                                '\n' : ' ');
                    }
                }
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private static void bkt(int pos, ArrayList<ArrayList<Integer>> all, ArrayList<Integer> crt_sol, ArrayList<Boolean> visited){
            if (pos == k){
                all.add(crt_sol);
                return;
            }

            for (int i = 1; i <= n; i++){
                if (!visited.get(i)){
                    visited.get(i) = true;
                    crt_sol.add(i);

                    bkt(pos + 1, all, crt_sol, visited);

                    visited[i] = false;
                    crt.sol = ArrayUtils.remove(crt_sol, crt_sol.size() - 1);
                }
            }
        }

        private ArrayList<ArrayList<Integer>> getResult() {

            ArrayList<ArrayList<Integer>> all = new ArrayList<>();
            ArrayList<Integer> crt_sol = new ArrayList<>();
            ArrayList<Boolean> visited = new ArrayList<>(n);
            Arrays.fill(visited, false);

            bkt(0, all, crt_sol, visited);

            return all;
        }

        public void solve() {
            readInput();
            writeOutput(getResult());
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}

File: LabPA/Lab5/cpp/task-1/main.cpp
#include <bits/stdc++.h>
using namespace std;

class Task {
 public:
    void solve() {
        read_input();
        print_output(get_result());
    }

 private:
    int n, k;

    void read_input() {
        ifstream fin("in");
        fin >> n >> k;
        fin.close();
    }

    void bkt(int pos, vector<vector<int>> &all, vector<int> &crt_sol, vector<bool> &visited){
        if (pos == k){
            all.emplace_back(crt_sol);
            return;
        }

        for (int i = 1; i <= n; i++){
            if (!visited[i]){
                visited[i] = true;
                crt_sol.emplace_back(i);

                bkt(pos + 1, all, crt_sol, visited);

                visited[i] = false;
                crt_sol.pop_back();
            }
        }
    }

    vector<vector<int> > get_result() {
        vector<vector<int> > all;
        vector<int> crt_sol;
        vector<bool> visited(n, false);

        bkt(0, all, crt_sol, visited);

        return all;
    }

    void print_output(vector<vector<int> > result) {
        ofstream fout("out");
        fout << result.size() << '\n';
        for (int i = 0; i < (int)result.size(); i++) {
            for (int j = 0; j < (int)result[i].size(); j++) {
                fout << result[i][j] <<
                    (j + 1 != result[i].size() ? ' ' : '\n');
            }
        }
        fout.close();
    }
};

// Please always keep this simple main function!
int main() {
    // Allocate a Task object on heap in order to be able to
    // declare huge static-allocated data structures inside the class.
    Task *task = new Task();
    task->solve();
    delete task;
    return 0;
}
